### ISSUE TITLE
Strange suggestion from suspicious_operation_groupings

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-clippy#6747]
Lint name: suspicious_operation_groupings

I tried this code:

```rust
    pub fn set_scheme(&mut self, scheme: &str) -> Result<(), ()> {
        let mut parser = Parser::for_setter(String::new());
        let remaining = parser.parse_scheme(parser::Input::new(scheme))?;
        let new_scheme_type = SchemeType::from(&parser.serialization);
        let old_scheme_type = SchemeType::from(self.scheme());
        // If url’s scheme is a special scheme and buffer is not a special scheme, then return.
        if (new_scheme_type.is_special() && !old_scheme_type.is_special()) ||
            // If url’s scheme is not a special scheme and buffer is a special scheme, then return.
            (!new_scheme_type.is_special() && old_scheme_type.is_special()) ||
            // If url includes credentials or has a non-null port, and buffer is "file", then return.
            // If url’s scheme is "file" and its host is an empty host or null, then return.
            (new_scheme_type.is_file() && self.has_authority())
        {
            return Err(());
        }
```

I expected to see this happen: accepted without warnings.

Instead, this happened:

```rust
error: This sequence of operators looks suspiciously like a bug.
Error:     --> url/src/lib.rs:2098:14
     |
2098 |             (new_scheme_type.is_file() && self.has_authority())
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: I think you meant: `new_scheme_type.is_file() && self.is_file()`
     |
     = note: `-D clippy::suspicious-operation-groupings` implied by `-D warnings`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_operation_groupings
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cb75ad5d 2021-02-10)
- `rustc -Vv`:
  ```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-apple-darwin
release: 1.50.0
  ```

**Comment by peterjoel** [comment]
I'm getting these all over the place. None of them are valid. e.g.:

```
warning: This sequence of operators looks suspiciously like a bug.
  --> stop_orders/src/lib.rs:68:32
   |
68 | ...                   && order.user_request_id == cancel_request.user_request_id_to_cancel
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: I think you meant: `order.user_request_id == cancel_request.user_request_id`
   |
   = note: `#[warn(clippy::suspicious_operation_groupings)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_operation_groupings
```

**Comment by camsteffen** [comment]
Similar to #6722

**Comment by michaelkirk** [comment]
Similar false positive:

```
warning: This sequence of operators looks suspiciously like a bug.
   --> geo/src/algorithm/line_intersection.rs:121:34
    |
121 |         if p.start == q.start || p.start == q.end {
    |                                  ^^^^^^^^^^^^^^^^ help: I think you meant: `p.end == q.end`
```

Some context:
```
if p.start == q.start || p.start == q.end {
    intersection = p.start;
} else if p.end == q.start || p.end == q.end {
    intersection = p.end;
    // Now check to see if any endpoint lies on the interior of the other segment.
} else if p_q1 == Collinear {
    intersection = q.start;
} else if p_q2 == Collinear {
    intersection = q.end;
} else if q_p1 == Collinear {
    intersection = p.start;
} else {
    assert_eq!(q_p2, Collinear);
    intersection = p.end;
}
```

I understand how this kind of a lint could find bugs, but it seems like it would also be unavoidably prone to false positives.

**Comment by Licenser** [comment]
I'd like to throw in a case here:
```
235 | /                 pc1.ast_eq(pc2)
236 | |                     && rest1.ast_eq(rest2)
    | |__________________________________________^ help: I think you meant: `pc1.ast_eq(pc2) && rest1.ast_eq(pc2)`
```
`rest1` and `pc1` are quite far apart spelling-wise (one haveing two more characters and not a single similar one). They also have different types, and the resulting code would not compile
